### PR TITLE
[IOTDB-1517][IOTDB-1521] Refactor TsFile Index for Vector(multi-variable timeseries)

### PR DESCRIPTION
**Pull request by Cpaulyz**
### TsFileSketchTool

Add new features.

* for*MetadataIndex*, show index tree to facilitate TsFile debug. 

	e.g.

	```
	|   [MetadataIndex:LEAF_DEVICE]
	|   └───────[root.sg_1.d1,2485]
	|                   [MetadataIndex:LEAF_MEASUREMENT]
	|                   └───────[s2,1742]
	|                   └───────[s6,1955]
	|                   └───────[vector.s2,2139]
	|   └───────[root.sg_1.d2,2535]
	|                   [MetadataIndex:LEAF_MEASUREMENT]
	|                   └───────[s2,2201]
	|                   └───────[s6,2414]
	```

* for *ChunkMatadataList*, show device info.

	* old version: `vector.s1`

	* new version: ` (root.sg_1.d1.)`

* for *Chunk*, show more message such as chunk header, page header.

### MetadataIndexConstructor

* **Construct**

	check according to [design document in confluence](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=184617773#TsFile%E6%96%87%E4%BB%B6%E6%A0%BC%E5%BC%8F-1.2.3.3IndexOfTimeseriesIndex%E6%97%B6%E9%97%B4%E5%BA%8F%E5%88%97%E7%B4%A2%E5%BC%95%E7%9A%84%E7%B4%A2%E5%BC%95%EF%BC%88%E4%BA%8C%E7%BA%A7%E7%B4%A2%E5%BC%95%EF%BC%89) example 5-7

* **Test**

	* add test about metadata index tree in `MetadataIndexConstructorTest.java`

	* to add more test case, you only need to create params `devices, vectorMeasurement, measurements` and call funtion `private void test(String[] devices, int[][] vectorMeasurement, String[][] singleMeasurement)` 

	* `test` funtion will automatically test by the following steps. Note that if the metadata index tree is re-designed, you may need to modify "generate correct result" in `generateCorrectResult`.

	* ```
		// 1. generate file
		// 2. read metadata from file
		// 3. generate correct result
		// 4. compare correct result with TsFile's metadata
		// 4.1 make sure device in order
		// 4.2 make sure timeseries in order
		// 4.3 make sure split leaf correctly
		```

### Read

* **motivation:** New design of vector structure may split its value entry into different leaf node.

* **Problem:**  However, query process reads subSensors accoring to the vector. As a result, in `TimeseriesMetadataCache.java`, funtion 

	```
	public List<TimeseriesMetadata> get(
	    TimeSeriesMetadataCacheKey key,
	    List<String> subSensorList,
	    Set<String> allSensors,
	    boolean debug)
	```

	read TimeseriesMetadataList by vector.

	For example, there is a vector `v1.(s1,s2)`, one leaf node stores [..., v1, v1.s1] and next node stores [v1.s2, ...]. Old implemtation call `reader.readTimeseriesMetadata(path, allSensors)`, which only read the leaf where time column stored. It may ignore the entry stored in other leaves such as v1.s2's entry.

* **Solution:** Extraction a common function `readTimeseriesMetadataForVector`, which reads the all timeseriesMetadata cover query key and subSensorList.

### CODE REVIEW

File: tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/MetadataIndexConstructor.java [hunk 1]
```diff
@@ -62,49 +68,45 @@ public static MetadataIndexNode constructMetadataIndex(
...
       int serializedTimeseriesMetadataNum = 0;
+      String vectorName = ""; // record previous vector name
+      int numOfValueColumns = 0;
       for (int i = 0; i < entry.getValue().size(); i++) {
         timeseriesMetadata = entry.getValue().get(i);
-        if (timeseriesMetadata.isTimeColumn()) {
-          // calculate the number of value columns in this vector
-          int numOfValueColumns = 0;
+        if (numOfValueColumns > 0) {
+          // must be value column，不清楚这里是否需要加一层检验？外层可否保证？
```
Comment (wangchao316) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/MetadataIndexConstructor.java:76]: Good  job.Thanks for your contribution.
I have a little question.
generate , chinese is used in  notes

File: tsfile/src/test/java/org/apache/iotdb/tsfile/write/MetadataIndexConstructorTest.java [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+      vectorMeasurement[i] = new int[0];
+      measurements[i] = new String[measurementNum];
+      for (int j = 0; j < measurementNum; j++) {
+        measurements[i][j] = measurementPrefix + generateIndexString(j, measurementNum);
+      }
+    }
+    test(devices, vectorMeasurement, measurements);
+  }
+
+  /** 150个实体，每个实体有1个物理量（docs.例3） */
```
Comment (wangchao316) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/write/MetadataIndexConstructorTest.java:118]: the same

File: tsfile/src/test/java/org/apache/iotdb/tsfile/write/MetadataIndexConstructorTest.java [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+        node.getNodeType().equals(MetadataIndexNodeType.LEAF_MEASUREMENT)
+            || node.getNodeType().equals(MetadataIndexNodeType.INTERNAL_MEASUREMENT));
+    for (int i = 0; i < node.getChildren().size(); i++) {
+      MetadataIndexEntry metadataIndexEntry = node.getChildren().get(i);
+      long endOffset = node.getEndOffset();
+      if (i != node.getChildren().size() - 1) {
+        endOffset = node.getChildren().get(i + 1).getOffset();
+      }
+      if (node.getNodeType().equals(MetadataIndexNodeType.LEAF_MEASUREMENT)) {
+        // 把每个叶子节点的第一个加进来
```
Comment (wangchao316) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/write/MetadataIndexConstructorTest.java:302]: the same

**Review by wangchao316 — CHANGES_REQUESTED** [review]
This pr changes the underlying data files too much, and we need to consider compatibility and performance. please use the benchmark to test whether the performance deteriorates.
Thanks.

File: session/src/test/java/org/apache/iotdb/session/IoTDBSessionVectorIT.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+import org.apache.iotdb.db.utils.EnvironmentUtils;
+import org.apache.iotdb.rpc.IoTDBConnectionException;
+import org.apache.iotdb.rpc.StatementExecutionException;
+import org.apache.iotdb.tsfile.common.conf.TSFileDescriptor;
+import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
+import org.apache.iotdb.tsfile.read.common.RowRecord;
+import org.apache.iotdb.tsfile.write.record.Tablet;
+import org.apache.iotdb.tsfile.write.schema.IMeasurementSchema;
+import org.apache.iotdb.tsfile.write.schema.VectorMeasurementSchema;
+
```
Comment (HTHou) [inline session/src/test/java/org/apache/iotdb/session/IoTDBSessionVectorIT.java:13]: Apache header....

**Comment by Cpaulyz** [comment]
rename TsFileSketchTool

* `ChunkMetadataList` -> `TimeseriesIndex`
* `ChunkMetadata` -> `ChunkIndex`

update document and fix some document bug, combine PR #3613

File: docs/zh/SystemDesign/TsFile/Format.md [hunk 5]
```diff
@@ -21,46 +21,74 @@
 
 # TsFile 文件格式
 
+[TOC]
```
Comment (SteveYurongSu) [inline docs/zh/SystemDesign/TsFile/Format.md:24]: Remove it~

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/TimeChunkWriter.java [hunk 6]
```diff
@@ -235,6 +235,7 @@ public void writeAllPagesOfChunkToTsFile(TsFileIOWriter writer) throws IOExcepti
 
     // write all pages of this column
     writer.writeBytesToStream(pageBuffer);
+    logger.debug("write time chunk {} from {} to {}", measurementId, dataOffset, writer.getPos());
```
Comment (SteveYurongSu) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/TimeChunkWriter.java:238]: Remove this debug log.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ValueChunkWriter.java [hunk 7]
```diff
@@ -234,6 +234,7 @@ public void writeAllPagesOfChunkToTsFile(TsFileIOWriter writer) throws IOExcepti
     writer.writeBytesToStream(pageBuffer);
 
     int dataSize = (int) (writer.getPos() - dataOffset);
+    logger.debug("write value chunk {} from {} to {}", measurementId, dataOffset, writer.getPos());
```
Comment (SteveYurongSu) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ValueChunkWriter.java:237]: Remove this debug log.

File: tsfile/src/test/java/org/apache/iotdb/tsfile/write/MetadataIndexConstructorTest.java [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+  /**
+   * Example 7: 2 entities, measurements of entities are shown in the following table
+   *
+   * <p>d0.s0~s4 | d0.v0.(s0~s8) | d0.z0~z3 d1.s0~s14 | d1.v0.(s0~s3)
+   */
+  @Test
+  public void compositeIndexTest() {
+    String[] devices = {"d0", "d1"};
+    int[][] vectorMeasurement = {{9}, {4}};
+    String[][] measurements = {
```
Comment (SteveYurongSu) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/write/MetadataIndexConstructorTest.java:181]: Rename it to singleMeasurement.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/MetadataIndexConstructor.java [hunk 9]
```diff
@@ -62,49 +68,45 @@ public static MetadataIndexNode constructMetadataIndex(
       MetadataIndexNode currentIndexNode =
           new MetadataIndexNode(MetadataIndexNodeType.LEAF_MEASUREMENT);
       int serializedTimeseriesMetadataNum = 0;
+      String vectorName = ""; // record previous vector name
```
Comment (SteveYurongSu) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/MetadataIndexConstructor.java:71]: Remove the field if it is not used.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/MetadataIndexConstructor.java [hunk 10]
```diff
@@ -62,49 +68,45 @@ public static MetadataIndexNode constructMetadataIndex(
...
+      int numOfValueColumns = 0;
       for (int i = 0; i < entry.getValue().size(); i++) {
         timeseriesMetadata = entry.getValue().get(i);
-        if (timeseriesMetadata.isTimeColumn()) {
-          // calculate the number of value columns in this vector
-          int numOfValueColumns = 0;
+        if (numOfValueColumns > 0) {
+          // must be value column
+          numOfValueColumns--;
+          timeseriesMetadata.setMeasurementId(timeseriesMetadata.getMeasurementId());
```
Comment (Cpaulyz) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/MetadataIndexConstructor.java:78]: useless

File: tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/MetadataIndexConstructor.java [hunk 11]
```diff
@@ -62,49 +68,45 @@ public static MetadataIndexNode constructMetadataIndex(
...
+        }
+        if (serializedTimeseriesMetadataNum == 0
+            || serializedTimeseriesMetadataNum >= config.getMaxDegreeOfIndexNode()) {
+          if (currentIndexNode.isFull()) {
+            addCurrentIndexNodeToQueue(currentIndexNode, measurementMetadataIndexQueue, out);
+            currentIndexNode = new MetadataIndexNode(MetadataIndexNodeType.LEAF_MEASUREMENT);
+          }
 
-          // for each vector, add time column of vector into LEAF_MEASUREMENT node
+          logger.debug(
```
Comment (Cpaulyz) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/MetadataIndexConstructor.java:96]: delete

File: tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/MetadataIndexConstructor.java [hunk 12]
```diff
@@ -120,28 +122,31 @@ public static MetadataIndexNode constructMetadataIndex(
           new MetadataIndexNode(MetadataIndexNodeType.LEAF_DEVICE);
       for (Map.Entry<String, MetadataIndexNode> entry : deviceMetadataIndexMap.entrySet()) {
         metadataIndexNode.addEntry(new MetadataIndexEntry(entry.getKey(), out.getPosition()));
+        logger.debug(
```
Comment (Cpaulyz) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/MetadataIndexConstructor.java:125]: delete

File: tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/MetadataIndexConstructor.java [hunk 13]
```diff
@@ -171,6 +176,8 @@ private static MetadataIndexNode generateRootNode(
         }
         currentIndexNode.addEntry(
             new MetadataIndexEntry(metadataIndexNode.peek().getName(), out.getPosition()));
+
```
Comment (Cpaulyz) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/MetadataIndexConstructor.java:179]: delete

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkGroupWriterImpl.java [hunk 14]
```diff
@@ -79,10 +87,72 @@ public void write(Tablet tablet) throws WriteProcessException {
...
+    // reference: MemTableFlushTask.java
+    int batchSize = tablet.rowSize;
+    VectorMeasurementSchema vectorMeasurementSchema =
+        (VectorMeasurementSchema) tablet.getSchemas().get(index);
+    List<TSDataType> valueDataTypes = vectorMeasurementSchema.getValueTSDataTypeList();
+    IChunkWriter vectorChunkWriter = chunkWriters.get(measurement);
+    for (int row = 0; row < batchSize; row++) {
+      long time = tablet.timestamps[row];
+      for (int columnIndex = 0; columnIndex < valueDataTypes.size(); columnIndex++) {
+        boolean isNull = false; // TODO: 这是干嘛用的？
```
Comment (Cpaulyz) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkGroupWriterImpl.java:118]: delete chinese

Comment (Cpaulyz) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkGroupWriterImpl.java:118]: isnull

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkWriterImpl.java [hunk 15]
```diff
@@ -440,6 +440,12 @@ private void writeAllPagesOfChunkToTsFile(TsFileIOWriter writer, Statistics<?> s
     writer.writeBytesToStream(pageBuffer);
 
     int dataSize = (int) (writer.getPos() - dataOffset);
+    logger.debug(
```
Comment (Cpaulyz) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkWriterImpl.java:443]: delete

File: tsfile/src/main/java/org/apache/iotdb/tsfile/write/writer/TsFileIOWriter.java [hunk 16]
```diff
@@ -354,6 +364,10 @@ private void flushOneChunkMetadata(
       if (!chunkMetadata.getDataType().equals(dataType)) {
         continue;
       }
+      logger.debug(
```
Comment (SteveYurongSu) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/write/writer/TsFileIOWriter.java:367]: remove

File: tsfile/src/test/java/org/apache/iotdb/tsfile/write/writer/VectorChunkWriterImplTest.java [hunk 17]
```diff
@@ -72,7 +74,7 @@ public void testWrite1() {
 
       // value chunk 1
       assertEquals(0x40 | MetaMarker.ONLY_ONE_PAGE_CHUNK_HEADER, ReadWriteIOUtils.readByte(buffer));
-      assertEquals("s1", ReadWriteIOUtils.readVarIntString(buffer));
+      assertEquals("s1.time.s1", ReadWriteIOUtils.readVarIntString(buffer));
```
Comment (Cpaulyz) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/write/writer/VectorChunkWriterImplTest.java:77]: rename -> vectorName

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java [hunk 18]
```diff
@@ -1266,6 +1276,23 @@ public long selfCheck(
     return res;
   }
 
+  /**
```
Comment (Cpaulyz) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java:1279]: remove to sketchTool

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java [hunk 19]
```diff
@@ -258,6 +258,12 @@ public byte readVersionNumber() throws IOException {
   public TsFileMetadata readFileMetadata() throws IOException {
     try {
       if (tsFileMetaData == null) {
+
```
Comment (Cpaulyz) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java:261]: delete

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java [hunk 20]
```diff
@@ -647,7 +653,11 @@ private void generateMetadataIndex(
             .computeIfAbsent(deviceId, k -> new ArrayList<>())
             .addAll(timeseriesMetadataList);
       } else {
-        deviceId = metadataIndex.getName();
+        // deviceId should be determined by LEAF_DEVICE node
+        if (type.equals(MetadataIndexNodeType.LEAF_DEVICE)) {
```
Comment (Cpaulyz) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java:657]: check

File: server/src/main/java/org/apache/iotdb/db/engine/cache/TimeSeriesMetadataCache.java [hunk 21]
```diff
@@ -251,6 +246,9 @@ public TimeseriesMetadata get(
       boolean debug)
       throws IOException {
     // put all sub sensors into allSensors
+    for (int i = 0; i < subSensorList.size(); i++) {
```
Comment (Cpaulyz) [inline server/src/main/java/org/apache/iotdb/db/engine/cache/TimeSeriesMetadataCache.java:249]: allSensors.removeAll(subSensorList); ?

**Review by SteveYurongSu — COMMENTED** [review]
Good job.

File: tsfile/src/test/java/org/apache/iotdb/tsfile/write/MetadataIndexConstructorTest.java [hunk 22]
```diff
@@ -0,0 +1,469 @@
...
+          // write Tablet to TsFile
+          if (tablet.rowSize != 0) {
+            tsFileWriter.write(tablet);
+            tablet.reset();
+          }
+        }
+      }
+      tsFileWriter.close();
+    } catch (Exception e) {
+      logger.error("meet error in TsFileWrite with tablet", e);
```
Comment (SteveYurongSu) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/write/MetadataIndexConstructorTest.java:450]:
```suggestion
      logger.error("meet error in TsFileWrite with tablet", e);
      fail(e.getMessage());
```

File: tsfile/src/test/java/org/apache/iotdb/tsfile/write/MetadataIndexConstructorTest.java [hunk 23]
```diff
@@ -0,0 +1,469 @@
...
+   */
+  private void generateFile(
+      String[] devices, int[][] vectorMeasurement, String[][] singleMeasurement) {
+    try {
+      File f = FSFactoryProducer.getFSFactory().getFile(FILE_PATH);
+      if (f.exists() && !f.delete()) {
+        throw new RuntimeException("can not delete " + f.getAbsolutePath());
+      }
+      Schema schema = new Schema();
+      TsFileWriter tsFileWriter = new TsFileWriter(f, schema);
```
Comment (SteveYurongSu) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/write/MetadataIndexConstructorTest.java:370]: Use try-with-resource...

File: tsfile/src/test/java/org/apache/iotdb/tsfile/write/MetadataIndexConstructorTest.java [hunk 24]
```diff
@@ -0,0 +1,469 @@
...
+   * @param devices load actual devices
+   * @param measurements load actual measurement(first of every leaf)
+   */
+  private void readMetaDataDFS(List<String> devices, List<List<String>> measurements) {
+    try (TsFileSequenceReader reader = new TsFileSequenceReader(FILE_PATH)) {
+      TsFileMetadata tsFileMetaData = reader.readFileMetadata();
+      MetadataIndexNode metadataIndexNode = tsFileMetaData.getMetadataIndex();
+      deviceDFS(devices, measurements, reader, metadataIndexNode);
+    } catch (IOException e) {
+      e.printStackTrace();
```
Comment (SteveYurongSu) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/write/MetadataIndexConstructorTest.java:256]:
```suggestion
      e.printStackTrace();
      fail(e.getMessage());
```

File: tsfile/src/test/java/org/apache/iotdb/tsfile/write/MetadataIndexConstructorTest.java [hunk 25]
```diff
@@ -0,0 +1,469 @@
...
+          reader.getAllTimeseriesMetadata();
+      for (int j = 0; j < actualDevices.size(); j++) {
+        for (int i = 0; i < actualMeasurements.get(j).size(); i++) {
+          assertEquals(
+              allTimeseriesMetadata.get(actualDevices.get(j)).get(i).getMeasurementId(),
+              correctFirstMeasurements.get(j).get(i));
+        }
+      }
+    } catch (IOException e) {
+      e.printStackTrace();
```
Comment (SteveYurongSu) [inline tsfile/src/test/java/org/apache/iotdb/tsfile/write/MetadataIndexConstructorTest.java:232]:
```suggestion
      e.printStackTrace();
      fail(e.getMessage());
```

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java [hunk 26]
```diff
@@ -647,7 +647,11 @@ private void generateMetadataIndex(
...
+          deviceId = metadataIndex.getName();
+        }
+        //        deviceId = metadataIndex.getName();
```
Comment (SteveYurongSu) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java:654]: remove it.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java [hunk 27]
```diff
@@ -1175,6 +1179,7 @@ public long selfCheck(
       }
       truncatedSize = this.position() - 1;
     } catch (Exception e) {
+      e.printStackTrace();
```
Comment (SteveYurongSu) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java:1182]: remove it.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java [hunk 28]
```diff
@@ -1266,6 +1271,23 @@ public long selfCheck(
...
   }
 
+  /**
+   * get metadata index node
+   *
+   * @param startOffset start read offset
+   * @param endOffset end read offset
+   * @return MetadataIndexNode
+   */
+  public MetadataIndexNode getMetadataIndexNode(long startOffset, long endOffset) {
```
Comment (SteveYurongSu) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/read/TsFileSequenceReader.java:1281]: Why can't we throw exceptions here?

File: session/src/test/java/org/apache/iotdb/session/IoTDBSessionVectorIT.java [hunk 29]
```diff
@@ -0,0 +1,219 @@
...
+        RowRecord rowRecord = dataSet.next();
+        assertEquals(rowRecord.getFields().get(0).getIntV(), time);
+        assertEquals(rowRecord.getFields().get(1).getIntV(), (time + 1) * 10 + 1);
+        assertEquals(rowRecord.getFields().get(2).getIntV(), (time + 1) * 10 + 2);
+        System.out.println(dataSet.next());
+      }
+
+      dataSet.closeOperationHandle();
+    } catch (IoTDBConnectionException e) {
+      e.printStackTrace();
```
Comment (SteveYurongSu) [inline session/src/test/java/org/apache/iotdb/session/IoTDBSessionVectorIT.java:86]: Let the test fail if an exception is thrown.

File: session/src/test/java/org/apache/iotdb/session/IoTDBSessionVectorIT.java [hunk 30]
```diff
@@ -0,0 +1,219 @@
...
+    Tablet tablet = new Tablet(ROOT_SG1_D1_VECTOR1, schemaList);
+    tablet.setAligned(true);
+    long timestamp = 0;
+
+    for (long row = 0; row < 100; row++) {
+      int rowIndex = tablet.rowSize++;
+      tablet.addTimestamp(rowIndex, timestamp);
+      tablet.addValue(
+          schemaList.get(0).getValueMeasurementIdList().get(0), rowIndex, row * 10 + 1L);
+      //          new SecureRandom().nextLong());
```
Comment (SteveYurongSu) [inline session/src/test/java/org/apache/iotdb/session/IoTDBSessionVectorIT.java:179]: Remove it?

File: tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/MetadataIndexConstructor.java [hunk 31]
```diff
@@ -64,47 +70,22 @@ public static MetadataIndexNode constructMetadataIndex(
...
-              currentIndexNode = new MetadataIndexNode(MetadataIndexNodeType.LEAF_MEASUREMENT);
-            }
-            currentIndexNode.addEntry(
-                new MetadataIndexEntry(timeseriesMetadata.getMeasurementId(), out.getPosition()));
-            serializedTimeseriesMetadataNum = 0;
-          }
-          timeseriesMetadata.serializeTo(out.wrapAsStream());
-          serializedTimeseriesMetadataNum++;
         }
+        logger.debug(
```
Comment (SteveYurongSu) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/MetadataIndexConstructor.java:83]: remove this line and the logger.

**Review by SteveYurongSu — CHANGES_REQUESTED** [review]

**Review by SteveYurongSu — APPROVED** [review]
LGTM.

File: docs/zh/SystemDesign/TsFile/Format.md [hunk 32]
```diff
@@ -27,39 +27,63 @@
...
+
+	- **1: DICTIONARY**
+
+		- 字典编码是一种无损编码。它适合编码基数小的数据（即数据去重后唯一值数量小）。不推荐用于基数大的数据。
+
+	- **2: RLE**
+
+		- 游程编码，比较适合存储某些整数值连续出现的序列，不适合编码大部分情况下前后值不一样的序列数据。
+
+		- 游程编码也可用于对浮点数进行编码，但在创建时间序列的时候需指定保留小数位数（MAX_POINT_NUMBER，具体指定方式参见本文[SQL 参考文档](https://iotdb.apache.org/zh/UserGuide/Master/Operation Manual/SQL Reference.html)）。比较适合存储某些浮点数值连续出现的序列数据，不适合存储对小数点后精度要求较高以及前后波动较大的序列数据。
```
Comment (SteveYurongSu) [inline docs/zh/SystemDesign/TsFile/Format.md:55]: Fix the dead link :D

**Review by SteveYurongSu — APPROVED** [review]

**Pull request merged**